Getting a reflection panic when trying to save a record with a `belongs_to` pointer

## Description

We are trying to update from `4.8.7` to `4.9.0`.

I have a model that uses pointers for relational fields that are optional (nullable.) For example:

```
type TransportationOffice struct {
    ID               uuid.UUID             `json:"id" db:"id"`                                             
    ShippingOfficeID *uuid.UUID            `json:"shipping_office_id" db:"shipping_office_id"`             
    ShippingOffice   *TransportationOffice `belongs_to:"transportation_offices"`
    Name             string                `json:"name" db:"name"`
    Address          Address               `belongs_to:"address"`
    ...
```
Full source [here](https://github.com/transcom/mymove/blob/master/pkg/models/transportation_office.go#L17)

Now, when trying to save this model with a null value in the `ShippingOffice` field, I'm getting a panic in the reflection package:

```
panic: reflect: call of reflect.Value.FieldByName on zero Value [recovered]
	panic: reflect: call of reflect.Value.FieldByName on zero Value

goroutine 66 [running]:
testing.tRunner.func1(0xc42058c000)
	/usr/local/Cellar/go@1.10/1.10.4/libexec/src/testing/testing.go:742 +0x29d
panic(0x18388a0, 0xc420671320)
	/usr/local/Cellar/go@1.10/1.10.4/libexec/src/runtime/panic.go:502 +0x229
reflect.flag.mustBe(0x0, 0x19)
	/usr/local/Cellar/go@1.10/1.10.4/libexec/src/reflect/value.go:207 +0xb5
reflect.Value.FieldByName(0x0, 0x0, 0x0, 0x191afb0, 0x2, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go@1.10/1.10.4/libexec/src/reflect/value.go:835 +0x4c
github.com/transcom/mymove/vendor/github.com/gobuffalo/pop/associations.(*belongsToAssociation).BeforeSetup(0xc42066e300, 0x185b940, 0x0)
	/Users/alexi/Documents/projects/golang/src/github.com/transcom/mymove/vendor/github.com/gobuffalo/pop/associations/belongs_to_association.go:113 +0x13e
github.com/transcom/mymove/vendor/github.com/gobuffalo/pop.(*Connection).Create.func1.1(0xbef2a84560ba7cae, 0x2a399c4b)
	/Users/alexi/Documents/projects/golang/src/github.com/transcom/mymove/vendor/github.com/gobuffalo/pop/executors.go:136 +0x8b7
github.com/transcom/mymove/vendor/github.com/gobuffalo/pop.(*Connection).timeFunc(0xc4200f53e0, 0x191eacb, 0x6, 0xc420d2b790, 0xc42046b680, 0xc420d2b7c8)
	/Users/alexi/Documents/projects/golang/src/github.com/transcom/mymove/vendor/github.com/gobuffalo/pop/connection.go:206 +0x50
github.com/transcom/mymove/vendor/github.com/gobuffalo/pop.(*Connection).Create.func1(0xc42052ba70, 0x23ab400, 0x0)
	/Users/alexi/Documents/projects/golang/src/github.com/transcom/mymove/vendor/github.com/gobuffalo/pop/executors.go:112 +0xa7
github.com/transcom/mymove/vendor/github.com/gobuffalo/pop.(*Model).iterate(0xc42052ba70, 0xc42052bad0, 0xc420d2b898, 0x156db9a)
	/Users/alexi/Documents/projects/golang/src/github.com/transcom/mymove/vendor/github.com/gobuffalo/pop/model.go:197 +0x20b
github.com/transcom/mymove/vendor/github.com/gobuffalo/pop.(*Connection).Create(0xc4200f53e0, 0x18fc7c0, 0xc42046b680, 0x0, 0x0, 0x0, 0x0, 0xc420d2b980)
	/Users/alexi/Documents/projects/golang/src/github.com/transcom/mymove/vendor/github.com/gobuffalo/pop/executors.go:111 +0xf0
github.com/transcom/mymove/vendor/github.com/gobuffalo/pop.(*Connection).ValidateAndCreate(0xc4200f53e0, 0x18fc7c0, 0xc42046b680, 0x0, 0x0, 0x0, 0x0, 0xc4200a4000, 0xc420d2b980)
	/Users/alexi/Documents/projects/golang/src/github.com/transcom/mymove/vendor/github.com/gobuffalo/pop/executors.go:100 +0x111
github.com/transcom/mymove/pkg/testdatagen.mustCreate(0xc4200f53e0, 0x18fc7c0, 0xc42046b680)
	/Users/alexi/Documents/projects/golang/src/github.com/transcom/mymove/pkg/testdatagen/shared.go:70 +0x6b
github.com/transcom/mymove/pkg/testdatagen.MakeTransportationOffice(0xc4200f53e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/alexi/Documents/projects/golang/src/github.com/transcom/mymove/pkg/testdatagen/make_transportation_office.go:24 +0x21d
```

It looks like some work was done in/around the `belongs_to_association.go` code recently: https://github.com/gobuffalo/pop/pull/199 that may be related.

Happy to provide more details if necessary.

--- a/associations/association.go
+++ b/associations/association.go
@@ -152,6 +152,9 @@ func fieldIsNil(f reflect.Value) bool {
 	if n := nulls.New(f.Interface()); n != nil {
 		return n.Interface() == nil
 	}
+	if f.Kind() == reflect.Interface || f.Kind() == reflect.Ptr {
+		return f.IsNil()
+	}
 	return f.Interface() == nil
 }
 

--- a/associations/belongs_to_association.go
+++ b/associations/belongs_to_association.go
@@ -93,19 +93,21 @@ func (b *belongsToAssociation) Constraint() (string, []interface{}) {
 }
 
 func (b *belongsToAssociation) BeforeInterface() interface{} {
+	// if the owner field is set, don't try to create the association to prevent conflicts.
 	if !b.skipped {
 		return nil
 	}
 
-	if b.ownerModel.Kind() == reflect.Ptr {
-		return b.ownerModel.Interface()
+	m := b.ownerModel
+	if m.Kind() == reflect.Ptr {
+		m = b.ownerModel.Elem()
 	}
 
-	if IsZeroOfUnderlyingType(b.ownerModel.Interface()) {
+	if IsZeroOfUnderlyingType(m.Interface()) {
 		return nil
 	}
 
-	return b.ownerModel.Addr().Interface()
+	return m.Addr().Interface()
 }
 
 func (b *belongsToAssociation) BeforeSetup() error {

--- a/associations/has_one_association.go
+++ b/associations/has_one_association.go
@@ -75,24 +75,12 @@ func (h *hasOneAssociation) Constraint() (string, []interface{}) {
 	return fmt.Sprintf("%s = ?", h.fkID), []interface{}{h.ownerID}
 }
 
-// AfterInterface gets the value of the model to create after
-// creating the parent model. It returns nil if its value is
-// not set.
-func (h *hasOneAssociation) AfterInterface() interface{} {
-	if h.ownedModel.Kind() == reflect.Ptr {
-		return h.ownedModel.Interface()
-	}
-
-	if IsZeroOfUnderlyingType(h.ownedModel.Interface()) {
+func (h *hasOneAssociation) AfterSetup() error {
+	om := h.ownedModel
+	if fieldIsNil(om) {
 		return nil
 	}
-
-	return h.ownedModel.Addr().Interface()
-}
-
-func (h *hasOneAssociation) AfterSetup() error {
 	ownerID := reflect.Indirect(reflect.ValueOf(h.owner)).FieldByName("ID").Interface()
-	om := h.ownedModel
 	if om.Kind() == reflect.Ptr {
 		om = om.Elem()
 	}
@@ -107,6 +95,24 @@ func (h *hasOneAssociation) AfterSetup() error {
 	}
 
 	return fmt.Errorf("could not set '%s' to '%s'", ownerID, fval)
+}
+
+// AfterInterface gets the value of the model to create after
+// creating the parent model. It returns nil if its value is
+// not set.
+func (h *hasOneAssociation) AfterInterface() interface{} {
+	m := h.ownedModel
+	if fieldIsNil(m) {
+		return nil
+	}
+	if m.Kind() == reflect.Ptr {
+		return m.Interface()
+	}
+	if IsZeroOfUnderlyingType(m.Interface()) {
+		return nil
+	}
+
+	return m.Addr().Interface()
 }
 
 func (h *hasOneAssociation) AfterProcess() AssociationStatement {

--- a/executors_test.go
+++ b/executors_test.go
@@ -916,15 +916,34 @@ func Test_Eager_Create_Belongs_To(t *testing.T) {
 func Test_Eager_Create_Belongs_To_Pointers(t *testing.T) {
 	transaction(func(tx *Connection) {
 		r := require.New(t)
+		// Create a body with a head
 		body := Body{
 			Head: &Head{},
 		}
 
 		err := tx.Eager().Create(&body)
 		r.NoError(err)
+		r.NotZero(body.ID)
+		r.NotZero(body.Head.ID)
 
 		ctx, _ := tx.Count(&Body{})
 		r.Equal(1, ctx)
+
+		ctx, _ = tx.Count(&Head{})
+		r.Equal(1, ctx)
+
+		// Create a body without a head:
+		body = Body{
+			Head: nil,
+		}
+
+		err = tx.Eager().Create(&body)
+		r.NoError(err)
+		r.NotZero(body.ID)
+		r.Nil(body.Head)
+
+		ctx, _ = tx.Count(&Body{})
+		r.Equal(2, ctx)
 
 		ctx, _ = tx.Count(&Head{})
 		r.Equal(1, ctx)

